File: backend/src/deployment.js
import {
  existsSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  statSync,
  writeFileSync,
} from "fs";
import { join, relative } from "path";
import { settings } from "./config.js";

const ACTION_ALIASES = {
  ajout: "ajout",
  add: "ajout",
  a: "ajout",
  suppression: "suppression",
  suppr: "suppression",
  delete: "suppression",
  del: "suppression",
  modification: "modification",
  modif: "modification",
  mod: "modification",
  modify: "modification",
};

const LINE_PATTERN = /^\s*(\S+)\s+(\S+)\s+\[([^\]]+)\]\s*$/;

function targetDir() {
  return join(settings.repoMountPath, settings.targetFolder);
}

function iterTargetFiles() {
  const base = targetDir();
  const files = [];
  if (!existsSync(base)) return files;
  function walk(dir) {
    for (const name of readdirSync(dir)) {
      const p = join(dir, name);
      if (statSync(p).isFile()) files.push(p);
      else if (statSync(p).isDirectory()) walk(p);
    }
  }
  walk(base);
  return files;
}

export function parseDeploymentLines(text) {
  const lines = [];
  const errors = [];
  text.split("\n").forEach((raw, idx) => {
    const stripped = raw.trim();
    const lineNum = idx + 1;
    if (!stripped || stripped.startsWith("#")) return;
    const match = stripped.match(LINE_PATTERN);
    if (!match) {
      errors.push(
        `Ligne ${lineNum}: format invalide — attendu '/source /cible [ACTION]'`,
      );
      return;
    }
    const [, source, target, actionRaw] = match;
    const actionKey = actionRaw.trim().toLowerCase();
    if (!ACTION_ALIASES[actionKey]) {
      errors.push(`Ligne ${lineNum}: action inconnue '${actionRaw}'`);
      return;
    }
    lines.push({
      line_number: lineNum,
      source,
      target,
      action: ACTION_ALIASES[actionKey],
      raw: stripped,
    });
  });
  return { lines, errors };
}

function redirectExists(source, target = null) {
  const repoRoot = settings.repoMountPath;
  const hits = [];
  for (const filePath of iterTargetFiles()) {
    let content;
    try {
      content = readFileSync(filePath, "utf-8");
    } catch {
      continue;
    }
    content.split("\n").forEach((line, idx) => {
      const stripped = line.trim();
      if (!stripped || stripped.startsWith("#")) return;
      const parts = stripped.split(/\s+/);
      if (!parts.includes(source)) return;
      if (target !== null && !parts.includes(target)) return;
      hits.push(`${relative(repoRoot, filePath)}:${idx + 1} — ${stripped}`);
    });
  }
  return hits;
}

export function buildPreview(deploymentLines) {
  const items = deploymentLines.map((line) => {
    if (line.action === "ajout") {
      const existing = redirectExists(line.source);
      if (existing.length) {
        return {
          line_number: line.line_number,
          source: line.source,
          target: line.target,
          action: line.action,
          status: "conflict",
          message: `Conflit: ${existing[0]}`,
          applicable: false,
        };
      }
      return {
        line_number: line.line_number,
        source: line.source,
        target: line.target,
        action: line.action,
        status: "ready",
        message: "Aucune occurrence existante",
        applicable: true,
      };
    }
    const existing = redirectExists(line.source);
    if (!existing.length) {
      return {
        line_number: line.line_number,
        source: line.source,
        target: line.target,
        action: line.action,
        status: "not_found",
        message: "Source introuvable",
        applicable: false,
      };
    }
    return {
      line_number: line.line_number,
      source: line.source,
      target: line.target,
      action: line.action,
      status: "ready",
      message: existing[0],
      applicable: true,
    };
  });

  return {
    items,
    applicable_count: items.filter((i) => i.applicable).length,
    conflict_count: items.filter((i) => i.status === "conflict").length,
    error_count: items.filter((i) => i.status === "not_found").length,
  };
}

function defaultRedirectsFile() {
  const base = targetDir();
  mkdirSync(base, { recursive: true });
  const htaccess = join(base, ".htaccess");
  if (existsSync(htaccess)) return htaccess;
  const confFiles = readdirSync(base).filter((f) => f.endsWith(".conf"));
  if (confFiles.length) return join(base, confFiles[0]);
  writeFileSync(htaccess, "# Redirections Apache\n", "utf-8");
  return htaccess;
}

export function applyDeployment(deploymentLines, selectedLineNumbers) {
  const repoRoot = settings.repoMountPath;
  const targetFile = defaultRedirectsFile();
  let contentLines = readFileSync(targetFile, "utf-8").split("\n");
  if (contentLines[contentLines.length - 1] === "") contentLines.pop();

  let adds = 0;
  let mods = 0;
  const modifiedFiles = new Set();

  for (const line of deploymentLines) {
    if (!selectedLineNumbers.has(line.line_number)) continue;
    if (line.action === "ajout") {
      contentLines.push(`Redirect 301 ${line.source} ${line.target}`);
      adds++;
      modifiedFiles.add(relative(repoRoot, targetFile));
    } else if (line.action === "suppression") {
      let removed = false;
      contentLines = contentLines.filter((existing) => {
        if (!removed && existing.includes(line.source)) {
          removed = true;
          return false;
        }
        return true;
      });
      modifiedFiles.add(relative(repoRoot, targetFile));
    } else if (line.action === "modification") {
      contentLines = contentLines.map((existing) => {
        if (
          existing.includes(line.source) &&
          (existing.includes("Redirect") || existing.includes("RewriteRule"))
        ) {
          return `Redirect 301 ${line.source} ${line.target}`;
        }
        return existing;
      });
      mods++;
      modifiedFiles.add(relative(repoRoot, targetFile));
    }
  }

  writeFileSync(targetFile, contentLines.join("\n") + "\n", "utf-8");
  return { adds, mods, files: [...modifiedFiles].sort() };
}

export function buildCommitMessage(preview) {
  const adds = preview.items.filter(
    (i) => i.applicable && i.action === "ajout",
  ).length;
  const suppr = preview.items.filter(
    (i) => i.applicable && i.action === "suppression",
  ).length;
  const mods = preview.items.filter(
    (i) => i.applicable && i.action === "modification",
  ).length;
  const parts = [];
  if (adds) parts.push(`${adds} ajout${adds > 1 ? "s" : ""}`);
  if (suppr) parts.push(`${suppr} suppression${suppr > 1 ? "s" : ""}`);
  if (mods) parts.push(`${mods} modification${mods > 1 ? "s" : ""}`);
  return `chore(redirects): ${parts.length ? parts.join(", ") : "mise à jour"}`;
}

export function exportDeploymentReport(data) {
  const lines = [
    "# Rapport déploiement",
    `Branche: ${data.branch}`,
    `Commit: ${data.commitSha || "—"}`,
    `Push: ${data.pushOk ? "OK" : "—"}`,
    `Ajouts: ${data.adds}`,
    `Suppressions: ${data.suppressions}`,
    `Modifications: ${data.modifications}`,
    "Fichiers:",
  ];
  for (const f of data.files) lines.push(`  - ${f}`);
  return lines.join("\n");
}

File: backend/app/deployment.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path

from app.config import Settings
from app.extraction import iter_target_files, search_url_in_file


class ActionType(str, Enum):
    AJOUT = "ajout"
    SUPPRESSION = "suppression"
    MODIFICATION = "modification"


ACTION_ALIASES: dict[str, ActionType] = {
    "ajout": ActionType.AJOUT,
    "add": ActionType.AJOUT,
    "a": ActionType.AJOUT,
    "suppression": ActionType.SUPPRESSION,
    "suppr": ActionType.SUPPRESSION,
    "delete": ActionType.SUPPRESSION,
    "del": ActionType.SUPPRESSION,
    "modification": ActionType.MODIFICATION,
    "modif": ActionType.MODIFICATION,
    "mod": ActionType.MODIFICATION,
    "modify": ActionType.MODIFICATION,
}


class PreviewStatus(str, Enum):
    READY = "ready"
    CONFLICT = "conflict"
    NOT_FOUND = "not_found"
    INVALID = "invalid"


@dataclass
class DeploymentLine:
    line_number: int
    source: str
    target: str
    action: ActionType
    raw: str


@dataclass
class PreviewItem:
    line_number: int
    source: str
    target: str
    action: ActionType
    status: PreviewStatus
    message: str
    applicable: bool = True


@dataclass
class DeploymentPreview:
    items: list[PreviewItem]
    applicable_count: int
    conflict_count: int
    error_count: int


LINE_PATTERN = re.compile(
    r"^\s*(\S+)\s+(\S+)\s+\[([^\]]+)\]\s*$",
)


def _normalize_action(action_raw: str) -> str:
    return action_raw.strip().lower()


def parse_deployment_lines(text: str) -> tuple[list[DeploymentLine], list[str]]:
    lines: list[DeploymentLine] = []
    errors: list[str] = []
    for idx, raw in enumerate(text.splitlines(), start=1):
        stripped = raw.strip()
        if not stripped or stripped.startswith("#"):
            continue
        match = LINE_PATTERN.match(stripped)
        if not match:
            errors.append(
                f"Ligne {idx}: format invalide — attendu '/source /cible [ACTION]'"
            )
            continue
        source, target, action_raw = match.groups()
        action_key = _normalize_action(action_raw)
        if action_key not in ACTION_ALIASES:
            errors.append(f"Ligne {idx}: action inconnue '{action_raw}'")
            continue
        lines.append(
            DeploymentLine(
                line_number=idx,
                source=source,
                target=target,
                action=ACTION_ALIASES[action_key],
                raw=stripped,
            )
        )
    return lines, errors


def redirect_exists(settings: Settings, source: str, target: str | None = None) -> list[str]:
    repo_root = Path(settings.repo_mount_path)
    hits: list[str] = []
    for file_path in iter_target_files(settings):
        try:
            lines = file_path.read_text(encoding="utf-8", errors="replace").splitlines()
        except OSError:
            continue
        for line_no, line in enumerate(lines, start=1):
            stripped = line.strip()
            if not stripped or stripped.startswith("#"):
                continue
            parts = stripped.split()
            if source not in parts:
                continue
            if target is not None and target not in parts:
                continue
            rel = str(file_path.relative_to(repo_root))
            hits.append(f"{rel}:{line_no} — {stripped}")
    return hits


def build_preview(settings: Settings, lines: list[DeploymentLine]) -> DeploymentPreview:
    items: list[PreviewItem] = []
    for line in lines:
        if line.action == ActionType.AJOUT:
            existing = redirect_exists(settings, line.source)
            if existing:
                items.append(
                    PreviewItem(
                        line_number=line.line_number,
                        source=line.source,
                        target=line.target,
                        action=line.action,
                        status=PreviewStatus.CONFLICT,
                        message=f"Conflit: {existing[0]}",
                        applicable=False,
                    )
                )
            else:
                items.append(
                    PreviewItem(
                        line_number=line.line_number,
                        source=line.source,
                        target=line.target,
                        action=line.action,
                        status=PreviewStatus.READY,
                        message="Aucune occurrence existante",
                        applicable=True,
                    )
                )
        else:
            existing = redirect_exists(settings, line.source)
            if not existing:
                items.append(
                    PreviewItem(
                        line_number=line.line_number,
                        source=line.source,
                        target=line.target,
                        action=line.action,
                        status=PreviewStatus.NOT_FOUND,
                        message="Source introuvable",
                        applicable=False,
                    )
                )
            else:
                items.append(
                    PreviewItem(
                        line_number=line.line_number,
                        source=line.source,
                        target=line.target,
                        action=line.action,
                        status=PreviewStatus.READY,
                        message=existing[0],
                        applicable=True,
                    )
                )

    applicable = sum(1 for i in items if i.applicable)
    conflicts = sum(1 for i in items if i.status == PreviewStatus.CONFLICT)
    errors = sum(1 for i in items if i.status == PreviewStatus.NOT_FOUND)
    return DeploymentPreview(
        items=items,
        applicable_count=applicable,
        conflict_count=conflicts,
        error_count=errors,
    )


def _default_redirects_file(settings: Settings) -> Path:
    base = Path(settings.repo_mount_path) / settings.target_folder
    base.mkdir(parents=True, exist_ok=True)
    htaccess = base / ".htaccess"
    if htaccess.exists():
        return htaccess
    conf_files = list(base.glob("*.conf"))
    if conf_files:
        return conf_files[0]
    htaccess.write_text("# Redirections Apache\n", encoding="utf-8")
    return htaccess


def apply_deployment(
    settings: Settings,
    lines: list[DeploymentLine],
    selected_line_numbers: set[int],
) -> tuple[int, int, list[str]]:
    modified_files: set[str] = set()
    adds = mods = 0
    repo_root = Path(settings.repo_mount_path)
    target_file = _default_redirects_file(settings)
    content_lines = target_file.read_text(encoding="utf-8", errors="replace").splitlines()

    for line in lines:
        if line.line_number not in selected_line_numbers:
            continue
        if line.action == ActionType.AJOUT:
            content_lines.append(f"Redirect 301 {line.source} {line.target}")
            adds += 1
            modified_files.add(str(target_file.relative_to(repo_root)))
        elif line.action == ActionType.SUPPRESSION:
            new_lines = []
            removed = False
            for existing in content_lines:
                if not removed and line.source in existing:
                    removed = True
                    continue
                new_lines.append(existing)
            content_lines = new_lines
            modified_files.add(str(target_file.relative_to(repo_root)))
        elif line.action == ActionType.MODIFICATION:
            new_lines = []
            for existing in content_lines:
                if line.source in existing and (
                    "Redirect" in existing or "RewriteRule" in existing
                ):
                    new_lines.append(f"Redirect 301 {line.source} {line.target}")
                else:
                    new_lines.append(existing)
            content_lines = new_lines
            mods += 1
            modified_files.add(str(target_file.relative_to(repo_root)))

    target_file.write_text("\n".join(content_lines) + "\n", encoding="utf-8")
    return adds, mods, sorted(modified_files)


def build_commit_message(preview: DeploymentPreview) -> str:
    adds = sum(1 for i in preview.items if i.applicable and i.action == ActionType.AJOUT)
    suppr = sum(
        1 for i in preview.items if i.applicable and i.action == ActionType.SUPPRESSION
    )
    mods = sum(
        1 for i in preview.items if i.applicable and i.action == ActionType.MODIFICATION
    )
    parts: list[str] = []
    if adds:
        parts.append(f"{adds} ajout{'s' if adds > 1 else ''}")
    if suppr:
        parts.append(f"{suppr} suppression{'s' if suppr > 1 else ''}")
    if mods:
        parts.append(f"{mods} modification{'s' if mods > 1 else ''}")
    detail = ", ".join(parts) if parts else "mise à jour"
    return f"chore(redirects): {detail}"


def export_deployment_report(
    branch: str,
    commit_sha: str | None,
    push_ok: bool,
    adds: int,
    suppr: int,
    mods: int,
    files: list[str],
) -> str:
    lines = [
        "# Rapport déploiement",
        f"Branche: {branch}",
        f"Commit: {commit_sha or '—'}",
        f"Push: {'OK' if push_ok else '—'}",
        f"Ajouts: {adds}",
        f"Suppressions: {suppr}",
        f"Modifications: {mods}",
        "Fichiers:",
    ]
    for f in files:
        lines.append(f"  - {f}")
    return "\n".join(lines)
